usr.sbin/atm/atm/atm.8
.\" @(#) atm.1,v 1.3 1997/05/05 17:25:27 johnc Exp
.de EX		\"Begin example
.ne 5
.if n .sp 1
.if t .sp .5
.nf
.in +.5i
..
.de EE
.fi
.in -.5i
.if n .sp 1
.if t .sp .5
..
.TH ATM 1 "1997-05-05" "HARP"
.SH NAME
atm \- user configuration and display command for ATM interface
.SH SYNOPSIS
Interface management subcommands:
.in +10
.ti -5
.B atm attach
<interface> <sigmgr>
.ti -5
.B atm detach
<interface>
.ti -5
.B atm set MAC
<interface> <MAC/ESI address>
.ti -5
.B atm set netif
<interface> <prefix> <count>
.ti -5
.B atm set prefix
<interface> <NSAP prefix>
.ti -5
.B atm show config
[<interface>]
.ti -5
.B atm show interface
[<interface>]
.ti -5
.B atm show netif
[<netif>]
.ti -5
.B atm show stats interface
[<interface> [phy | dev | atm | aal0 | aal4 | aal5 | driver]]
.sp
.ti -10
VCC management subcommands:
.ti -5
.B atm add PVC
<interface> <vpi> <vci> <aal> <encaps> <owner> ...
.ti -5
.B atm delete PVC
<interface> <vpi> <vci>
.ti -5
.B atm delete SVC
<interface> <vpi> <vci>
.ti -5
.B atm show stats VCC
[<interface> [<vpi> [<vci>]]]
.ti -5
.B atm show VCC
[<interface> [<vpi> [<vci>] | SVC | PVC]]
.sp
.ti -10
IP management subcommands:
.ti -5
.B atm add ARP
<host> <ATM address>
.ti -5
.B atm add PVC
<interface> <vpi> <vci> <aal> <encaps> IP <netif> <host> | dynamic
.ti -5
.B atm delete ARP
<host>
.ti -5
.B atm set arpserver
<netif> <ATM address> | local
.ti -5
.B atm show ARP
[<host>]
.ti -5
.B atm show arpserver
[<netif>]
.ti -5
.B atm show IPVCC
[<host> | <netif>]
.ti -5
.sp
.ti -10
Miscellaneous subcommands:
.ti -5
.B atm help
.ti -5
.B atm show version
.in -10
.fi
.SH DESCRIPTION
.I atm
configures and displays the status of the Host ATM Research Platform
(HARP) networking software.
The subcommands fall into several categories:
.PP
\fIInterface management\fP subcommands allow manipulation of the
ATM interface.
Functions include assigning a signalling manager to an interface,
setting the ATM address, associating network interfaces with
an interface, and displaying information about interfaces.
.PP
\fIVCC management\fP subcommands allow for managing ATM virtual
channel connections (VCCs).
Functions include opening and closing VCCs and displaying information
about them.
.PP
\fIIP management\fP subcommands allow for managing the interface
between IP and the ATM software.
Functions include displaying and manipulating the ARP cache,
opening a PVC connected to IP,
assigning an ARP server to a network interface,
and displaying information about IP VCCs.
.PP
\fIMiscellaneous\fP subcommands allow for displaying the version
of the ATM software and for getting help with the \fIatm\fP command.
.SS "Signalling Managers"
The signalling manager is responsible for the opening and closing of
VCCs.
Three signalling managers are supported:
.PP
.in +10
.ti -5
PVC - for PVCs only,
.ti -5
SPANS - supports SPANS, FORE's proprietary signalling protocol,
.ti -5
UNI 3.0 - supports the signalling protocol from The ATM Forum's
\fIATM User-Network Interface Specification, Version 3.0\fP.
.in -10
.PP
All three signalling managers support the opening and closing of PVCs
(see the \fIadd\fP and \fIdelete\fP subcommands).
.PP
A signalling manager must be attached to a physical interface
(see the \fIattach\fP subcommand)
before any VCCs can be created on the interface.
.SS "Physical and Network Interfaces"
Two types of interfaces are supported:
physical interfaces and network interfaces.
A physical interface represents a physical point of attachment to an
ATM network.
A physical interface has an ATM address associated with it.
.PP
A network interface is a logical interface.
One or more network interfaces are associated with a physical
interface;  each network interface has an IP address associated with it.
For SIGPVC- or SPANS-controlled interfaces, there must be one and
only one network interface associated with each physical interface.
For UNI-controlled interfaces, there can be up to 256 network
interfaces associated with a physical interface.
In this case, the correspondence between the network interface and
the ATM address is determined by the selector field (the last
byte) of the physical interface's ATM address.
.SS "Keyword and Documentation Conventions"
Command and subcommand keywords can be abbreviated by simply giving
enough of the first part of the keyword to make it unique.
Thus, \fIatm sh v\fB gives the same result as \fIatm show vcc\fB.
.PP
All keywords are case-insensitive.
.PP
Where a host address needs to be given to the \fIatm\fP command,
either a DNS name or an IP address in dotted decimal format can
be used.
.PP
ATM addresses are specified as strings of hex digits, with an
optional leading "0x".
Fields within the address may be separated by periods.
SPANS addresses are 8 bytes long, while NSAP-format addresses
are 20 bytes long.
The full address, including any leading zeroes, must be given.
For example:
.in +5
0x47.0005.80.ffe100.0000.f21a.0170.0020481a0170.00 (NSAP format)
.br
0x00000010.f2050aa9 (SPANS format)
.in -5
.fi
.SH SUBCOMMANDS
.SS Interface Management Subcommands:
.in +5
.ti -5
\fIatm add PVC <interface> <vpi> <vci> <aal> <encaps> <owner> ...\fP
.in -5
.PP
the format of the \fIadd PVC\fP subcommand varies depending on the
owner of the PVC.
See the description under "IP Management Subcommands."
.PP
\fIatm attach <interface> <sigmgr>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface to which the
signalling manager is to be attached,
.ti -5
\fI<sigmgr>\fP specifies which signalling manager is to be attached.
Valid choices are "SIGPVC", "SPANS", and "UNI30".
.in -10
.PP
This command attaches a signalling manager to an interface.
Until this is done, VCCs cannot be opened or closed.
Only one signalling manager at a time can be attached to an interface.
.PP
\fIatm detach <interface>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface whose signalling
manager is to be detached.
.in -10
.PP
This command detaches a signalling manager from an interface.
All VCCs that the signalling manager has created will be closed,
and no new VCCs can be created until a signalling manager (either
the same or a different one) is attached again.
.PP
\fIatm set MAC <interface> <MAC/ESI address>\fP
.PP
where:
.in +10
.ti -5
\fI<interface>\fP specifies the physical interface whose MAC address
is to be set,
.ti -5
\fI<MAC/ESI address>\fP specifies the MAC part of the NSAP address
for the interface.
The MAC address is specified as a string of hexadecimal digits with an
optional leading "0x".
Fields in the address may be separated by periods.
.in -10
.PP
This command sets the MAC address for a UNI-controlled interface.
The first 13 bytes (the prefix) of the 20-byte NSAP-format address
are set by the \fIset prefix\fP subcommand

usr.sbin/atm/atm/atm.c
/*
 *
 * ===================================
 * HARP  |  Host ATM Research Platform
 * ===================================
 *
 *
 * Copyright (c) 1994-1998, Network Computing Services, Inc.
 * All rights reserved.
 *
 *	@(#) atm.c,v 1.17 1998/08/06 16:57:47 johnc Exp
 *
 */

/*
 * User configuration and display program
 * --------------------------------------
 *
 * Main routine
 *
 */

#ifndef lint
static char *RCSid = "@(#) atm.c,v 1.17 1998/08/06 16:57:47 johnc Exp";
#endif

#include <netatm/user_include.h>

#include <stdio.h>
#include <errno.h>
#include <ctype.h>
#include <string.h>
#include <sys/ioctl.h>
#include "../lib/libatm.h"
#include "atm.h"


/*
 * Usage string
 */
#define USAGE_STR	"Interface management subcommands:\n\
    attach <intf> <protocol>\n\
    detach <intf>\n\
    set mac <intf> <MAC/ESI address>\n\
    set netif <intf> <prefix> <n>\n\
    set prefix <intf> <NSAP prefix>\n\
    show config [<intf>]\n\
    show interface [<intf>]\n\
    show netif [<netif>]\n\
    show stats interface [<intf> phy | dev | atm | aal0 | aal4 |\n\
        aal5 | driver]\n\
\n\
VC management subcommands:\n\
    add pvc <intf> <vpi> <vci> <aal> <encaps> <owner> ...\n\
    delete pvc <intf> <vpi> <vci>\n\
    delete svc <intf> <vpi> <vci>\n\
    show stats vcc [<intf> [vpi [vci]]]\n\
    show vcc [<intf> [<vpi> [<vci>] | SVC | PVC]]\n\
\n\
IP management subcommands:\n\
    add arp [<netif>] <IP addr> <ATM addr>\n\
    add pvc <intf> <vpi> <vci> <aal> <encaps> IP <netif> <IP addr> |\n\
        dynamic\n\
    delete arp [<netif>] <IP addr>\n\
    set arpserver <netif> <server> <IP prefix> ...\n\
    show arp [<host>]\n\
    show arpserver [<netif>]\n\
    show ipvcc [<IP addr> | <netif>]\n\
\n\
Miscellaneous subcommands:\n\
    help\n\
    show version\n"


/*
 * Local definitions
 */

struct cmd	add_subcmd[];
struct cmd	dlt_subcmd[];
struct cmd	set_subcmd[];
struct cmd	show_subcmd[];
struct cmd	stats_subcmd[];

struct cmd	cmds[] = {
	{ "add",	0,	0,	NULL,	(char *) add_subcmd },
	{ "attach",	2,	2,	attach,	"<intf> <protocol>" },
	{ "delete",	0,	0,	NULL,	(char *) dlt_subcmd },
	{ "detach",	1,	1,	detach,	"<intf>" },
	{ "set",	0,	0,	NULL,	(char *) set_subcmd },
	{ "show",	0,	0,	NULL,	(char *) show_subcmd },
	{ "help",	0,	99,	help,	"" },
	{ 0,		0,	0,	NULL,	"" }
};

struct cmd add_subcmd[] = {
	{ "arp",	2,	3,	arp_add, "[<netif>] <IP addr> <ATM addr>" },
	{ "pvc",	6,	12,	pvc_add, "<intf> <vpi> <vci> <aal> <encaps> <owner> ..." },
	{ 0,		0,	0,	NULL,	"" }
};

struct cmd dlt_subcmd[] = {
	{ "arp",	1,	2,	arp_dlt, "[<netif>] <IP addr>" },
	{ "pvc",	3,	3,	pvc_dlt, "<intf> <vpi> <vci>" },
	{ "svc",	3,	3,	svc_dlt, "<intf> <vpi> <vci>" },
	{ 0,		0,	0,	NULL,	"" }
};

struct cmd set_subcmd[] = {
	{ "arpserver",	2,	18,	set_arpserver, "<netif> <server>" },
	{ "mac",	2,	2,	set_macaddr, "<intf> <MAC/ESI address>" },
	{ "netif",	3,	3,	set_netif, "<intf> <prefix> <n>" },
	{ "prefix",	2,	2,	set_prefix, "<intf> <NSAP prefix>" },
	{ 0,		0,	0,	NULL,	""}
};

struct cmd show_subcmd[] = {
	{ "arp",	0,	1,	show_arp, "[<host>]" },
	{ "arpserver",	0,	1,	show_arpserv, "[<netif>]" },
	{ "config",	0,	1,	show_config, "[<intf>]" },
	{ "interface",	0,	1,	show_intf, "[<intf>]" },
	{ "ipvcc",	0,	3,	show_ip_vcc, "[<IP addr> | <netif>]" },
	{ "netif",	0,	1,	show_netif, "[<netif>]" },
	{ "stats",	0,	3,	NULL, (char *) stats_subcmd },
	{ "vcc",	0,	3,	show_vcc, "[<intf>] [<vpi> [<vci>] | SVC | PVC]" },
	{ "version",	0,	0,	show_version, "" },
	{ 0,		0,	0,	NULL,	"" }
};

struct cmd stats_subcmd[] = {
	{ "interface",	0,	2,	show_intf_stats, "[<intf> [cfg | phy | dev | atm | aal0 | aal4 | aal5 | driver]]" },
	{ "vcc",	0,	3,	show_vcc_stats, "[<intf> [vpi [vci]]]" },
	{ 0,		0,	0,	NULL,	"" }
};


/*
 * Supported signalling protocols
 */
struct proto	protos[] = {
	{ "SIGPVC",	ATM_SIG_PVC },
	{ "SPANS",	ATM_SIG_SPANS },
	{ "UNI30",	ATM_SIG_UNI30 },
	{ "UNI31",	ATM_SIG_UNI31 },
	{ "UNI40",	ATM_SIG_UNI40 },
	{ 0,		0 }
};

/*
 * Supported VCC owners
 */
struct owner	owners[] = {
	{ "IP",		ENDPT_IP,	ip_pvcadd },
	{ "SPANS",	ENDPT_SPANS_SIG,0 },
	{ "SPANS CLS",	ENDPT_SPANS_CLS,0 },
	{ "UNI SIG",	ENDPT_UNI_SIG,	0 },
	{ 0,		0,		0 }
};

/*
 * Supported AAL parameters
 */
struct aal	aals[] = {
	{ "Null",	ATM_AAL0 },
	{ "AAL0",	ATM_AAL0 },
	{ "AAL1",	ATM_AAL1 },
	{ "AAL2",	ATM_AAL2 },
	{ "AAL4",	ATM_AAL3_4 },
	{ "AAL3",	ATM_AAL3_4 },
	{ "AAL3/4",	ATM_AAL3_4 },
	{ "AAL5",	ATM_AAL5 },
	{ 0,	0 },
};

/*
 * Supported VCC encapsulations
 */
struct encaps	encaps[] = {
	{ "Null",	ATM_ENC_NULL },
	{ "None",	ATM_ENC_NULL },
	{ "LLC/SNAP",	ATM_ENC_LLC },
	{ "LLC",	ATM_ENC_LLC },
	{ "SNAP",	ATM_ENC_LLC },
	{ 0,	0 },
};


char	*prog;
char	prefix[128] = "";


main(argc, argv)
	int	argc;
	char	**argv;
{
	int	error;

	/*
	 * Save program name, ignoring any path components
	 */
	if (prog = (char *)strrchr(argv[0], '/'))
		prog++;
	else
		prog = argv[0];

	if (argc < 2) {
		usage(cmds, "");
		exit(1);
	}
	argc--; argv++;
		
	/*
	 * Validate and process command
	 */
	if (error = do_cmd(cmds, argc, argv))
		usage(cmds, "");

	exit(error);
}


/*
 * Validate and process user command
 * 
 * Arguments:
 *	descp	pointer to command description array
 *	argc	number of arguments left in command
 *	argv	pointer to argument strings
 *
 * Returns:
 *	none
 *
 */
int
do_cmd(descp, argc, argv)
	struct cmd	*descp;
	int		argc;
	char		**argv;
{
	struct cmd	*cmdp = 0;

	/*
	 * Make sure we have paramaters to process
	 */
	if (!argc) {
		usage(cmds, "");
		exit(1);
	}

	/*
	 * Figure out what command user wants
	 */
	for (; descp->name; descp++) {
		/*
		 * Use an exact match if there is one
		 */
		if (!strcasecmp(descp->name, argv[0])) {
			cmdp = descp;
			break;
		}
		/*
		 * Look for a match on the first part of keyword
		 */
		if (!strncasecmp(descp->name, argv[0], strlen(argv[0]))) {
			if (cmdp) {
				fprintf(stderr, "%s: Ambiguous parameter \"%s\"\n",
						prog, argv[0]);
				exit(1);
			}
			cmdp = descp;
		}
	}
	if (!cmdp)
		return(1);
	argc--; argv++;

	/*
	 * See if this command has subcommands
	 */
	if (cmdp->func == NULL) {
		strcat(prefix, cmdp->name);
		strcat(prefix, " ");
		return(do_cmd((struct cmd *)cmdp->help, argc, argv));
	}

	/*
	 * Minimal validation
	 */
	if ((argc < cmdp->minp) || (argc > cmdp->maxp)) {
		fprintf(stderr, "%s: Invalid number of arguments\n",
			prog);
		fprintf(stderr, "\tformat is: %s%s %s\n",
			prefix, cmdp->name, cmdp->help);
		exit(1);
	}

	/*
	 * Process command
	 */
	(*cmdp->func)(argc, argv, cmdp);
	return(0);
}


/*
 * Print command usage information
 * 
 * Arguments:
 *	cmdp	pointer to command description 
 * 	pref	pointer current command prefix 
 *
 * Returns:
 *	none
 *
 */
void
usage(cmdp, pref)
	struct cmd	*cmdp;
	char		*pref;
{
	fprintf(stderr, "usage: %s command [arg] [arg]...\n", prog);
	fprintf(stderr, USAGE_STR);
}


/*
 * Process interface attach command
 * 
 * Command format: 
 *	atm attach <interface_name> <protocol_name>
 *
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
attach(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	struct atmcfgreq	aar;
	struct proto	*prp;
	int		s;

	/*
	 * Validate interface name
	 */
	if (strlen(argv[0]) > sizeof(aar.acr_att_intf) - 1) {
		fprintf(stderr, "%s: Illegal interface name\n", prog);
		exit(1);
	}

	/*
	 * Find/validate requested signalling protocol
	 */
	for (prp = protos; prp->p_name; prp++) {
		if (strcasecmp(prp->p_name, argv[1]) == 0)
			break;
	}
	if (prp->p_name == NULL) {
		fprintf(stderr, "%s: Unknown signalling protocol\n", prog);
		exit(1);
	}


	/*
	 * Build ioctl request
	 */
	aar.acr_opcode = AIOCS_CFG_ATT;
	strncpy(aar.acr_att_intf, argv[0], sizeof(aar.acr_att_intf));
	aar.acr_att_proto = prp->p_id;

	/*
	 * Tell the kernel to do the attach
	 */
	s = socket(AF_ATM, SOCK_DGRAM, 0);
	if (s < 0) {
		sock_error(errno);
	}
	if (ioctl(s, AIOCCFG, (caddr_t)&aar) < 0) {
		fprintf(stderr, "%s: ", prog);
		switch (errno) {
		case EINVAL:
		case EOPNOTSUPP:
		case EPROTONOSUPPORT:
			perror("Internal error");
			break;
		case ENOMEM:
			fprintf(stderr, "Kernel memory exhausted\n");
			break;
		case EEXIST:
			fprintf(stderr, "Signalling manager already attached to %s\n",
					argv[0]);
			break;
		case ENETDOWN:
			fprintf(stderr, "ATM network is inoperable\n");
			break;
		case EPERM:
			fprintf(stderr, "Must be super user to use attach subcommand\n");
			break;
		case ENXIO:
			fprintf(stderr, "%s is not an ATM device\n",
					argv[0]);
			break;
		case ETOOMANYREFS:
			fprintf(stderr, "%s has too few or too many network interfaces\n",
					argv[0]);
			break;
		default:
			perror("Ioctl (AIOCCFG) attach");
			break;
		}
		exit(1);
	}
	(void)close(s);
}


/*
 * Process interface detach command
 * 
 * Command format: 
 *	atm detach <interface_name>
 *
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
detach(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	struct atmcfgreq	adr;
	int		s;

	/*
	 * Validate interface name
	 */
	if (strlen(argv[0]) > sizeof(adr.acr_det_intf) - 1) {
		fprintf(stderr, "%s: Illegal interface name\n", prog);
		exit(1);
	}

	/*
	 * Build ioctl request
	 */
	adr.acr_opcode = AIOCS_CFG_DET;
	strncpy(adr.acr_det_intf, argv[0], sizeof(adr.acr_det_intf));

	/*
	 * Tell the kernel to do the detach
	 */
	s = socket(AF_ATM, SOCK_DGRAM, 0);
	if (s < 0) {
		sock_error(errno);
	}
	if (ioctl(s, AIOCCFG, (caddr_t)&adr) < 0) {
		fprintf(stderr, "%s: ", prog);
		switch (errno) {
		case EALREADY:
			fprintf(stderr, "Signalling manager already detaching from %s\n",
					argv[0]);
			break;
		case EINVAL:
			perror("Internal error");
			break;
		case EPERM:
			fprintf(stderr, "Must be super user to use detach subcommand\n");
			break;
		default:
			perror("ioctl (AIOCCFG) detach");
			break;
		}
		exit(1);
	}
	(void)close(s);
}


/*
 * Process PVC add command
 * 
 * Command format: 
 *	atm add PVC <interface_name> <vpi> <vci> <aal> <encaps>
 *		<owner_name>
 *
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
pvc_add(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	struct atmaddreq	apr;
	struct atminfreq	air;
	struct air_int_rsp	*int_info;
	struct owner	*owp;
	struct aal	*alp;
	struct encaps	*enp;
	char	*cp;
	long	v;
	int	buf_len, s;

	/*
	 * Initialize opcode and flags
	 */
	apr.aar_opcode = AIOCS_ADD_PVC;
	apr.aar_pvc_flags = 0;

	/*
	 * Validate interface name and issue an information
	 * request IOCTL for the interface
	 */
	if (strlen(argv[0]) > sizeof(apr.aar_pvc_intf) - 1) {
		fprintf(stderr, "%s: Illegal interface name\n", prog);
		exit(1);
	}
	UM_ZERO(air.air_int_intf, sizeof(air.air_int_intf));
	strcpy(air.air_int_intf, argv[0]);
	buf_len = sizeof(struct air_int_rsp);
	air.air_opcode = AIOCS_INF_INT;
	buf_len = do_info_ioctl(&air, buf_len);
	if (buf_len < 0) {
		fprintf(stderr, "%s: ", prog);
		switch (errno) {
		case ENOPROTOOPT:
		case EOPNOTSUPP:
			perror("Internal error");
			break;
		case ENXIO:
			fprintf(stderr, "%s is not an ATM device\n",
					argv[0]);
			break;
		default:
			perror("ioctl (AIOCINFO)");
			break;
		}
		exit(1);
	}
	int_info = (struct air_int_rsp *) air.air_buf_addr;
	strcpy(apr.aar_pvc_intf, argv[0]);
	argc--; argv++;

	/*
	 * Validate vpi/vci values
	 */
	v = strtol(argv[0], &cp, 0);
	if ((*cp != '\0') || (v < 0) || (v >= 1 << 8)) {
		fprintf(stderr, "%s: Invalid VPI value\n", prog);
		exit(1);
	}
	apr.aar_pvc_vpi = (u_short) v;
	argc--; argv++;

	v = strtol(argv[0], &cp, 0);
	if ((*cp != '\0') || (v < MIN_VCI) || (v >= 1 << 16)) {
		fprintf(stderr, "%s: Invalid VCI value\n", prog);
		exit(1);
	}
	apr.aar_pvc_vci = (u_short) v;
	argc--; argv++;

	/*
	 * Validate requested PVC AAL
	 */
	for (alp = aals; alp->a_name; alp++) {
		if (strcasecmp(alp->a_name, argv[0]) == 0)
			break;
	}
	if (alp->a_name == NULL) {
		fprintf(stderr, "%s: Invalid PVC AAL\n", prog);
		exit(1);
	}
	apr.aar_pvc_aal = alp->a_id;
	argc--; argv++;

	/*
	 * Validate requested PVC encapsulation
	 */
	for (enp = encaps; enp->e_name; enp++) {
		if (strcasecmp(enp->e_name, argv[0]) == 0)
			break;
	}
	if (enp->e_name == NULL) {
		fprintf(stderr, "%s: Invalid PVC encapsulation\n", prog);
		exit(1);
	}
	apr.aar_pvc_encaps = enp->e_id;
	argc--; argv++;

	/*
	 * Validate requested PVC owner
	 */
	for (owp = owners; owp->o_name; owp++) {
		if (strcasecmp(owp->o_name, argv[0]) == 0)
			break;
	}
	if (owp->o_name == NULL) {
		fprintf(stderr, "%s: Unknown PVC owner\n", prog);
		exit(1);
	}
	apr.aar_pvc_sap = owp->o_sap;
	argc--; argv++;

	/*
	 * Perform service user processing
	 */
	if (owp->o_pvcadd) {
		(*owp->o_pvcadd)(argc, argv, cmdp, &apr, int_info);
	} else {
		fprintf(stderr, "%s: Unsupported PVC owner\n", prog);
		exit(1);
	}

	/*
	 * Tell the kernel to add the PVC
	 */
	s = socket(AF_ATM, SOCK_DGRAM, 0);
	if (s < 0) {
		sock_error(errno);
	}
	if (ioctl(s, AIOCADD, (caddr_t)&apr) < 0) {
		fprintf(stderr, "%s: ", prog);
		switch (errno) {
		case EPROTONOSUPPORT:
		case ENOPROTOOPT:
			perror("Internal error");
			break;
		case EINVAL:
			fprintf(stderr, "Invalid parameter\n");
			break;
		case EEXIST:
			fprintf(stderr, "PVC already exists\n");
			break;
		case ENETDOWN:
			fprintf(stderr, "ATM network is inoperable\n");
			break;
		case ENOMEM:
			fprintf(stderr, "Kernel memory exhausted\n");
			break;
		case EPERM:
			fprintf(stderr, "Must be super user to use add subcommand\n");
			break;
		case ERANGE:
			fprintf(stderr, "Invalid VPI or VCI value\n");
			break;
		default:
			perror("ioctl (AIOCADD) add PVC");
			break;
		}
		exit(1);
	}
	(void)close(s);
}


/*
 * Process ARP add command
 * 
 * Command formats: 
 *	atm add arp [<netif>] <IP addr> <ATM addr>
 *
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
arp_add(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	int			len, rc, s;
	struct atmaddreq	apr;
	Atm_addr		host_atm;
	struct sockaddr_in	*sin;
	union {
		struct sockaddr_in	sin;
		struct sockaddr		sa;
	} host_ip;

	/*
	 * Initialize add request structure
	 */
	UM_ZERO(&apr, sizeof(apr));

	/*
	 * Get network interface name if one is present
	 */
	if (argc == 3) {
		check_netif_name(argv[0]);
		strcpy(apr.aar_arp_intf, argv[0]);
		argc--; argv++;
	}

        /*
         * Get IP address of specified host name
         */
	UM_ZERO(&host_ip, sizeof(host_ip));
	host_ip.sa.sa_family = AF_INET;
	sin = get_ip_addr(argv[0]);
	host_ip.sin.sin_addr.s_addr = sin->sin_addr.s_addr;
	argc--; argv++;

	/*
	 * Get specified ATM address
	 */
	len = get_hex_atm_addr(argv[0], (u_char *)host_atm.address,
			sizeof(Atm_addr_nsap));
	switch(len) {
	case sizeof(Atm_addr_nsap):
		host_atm.address_format = T_ATM_ENDSYS_ADDR;
		host_atm.address_length = sizeof(Atm_addr_nsap);
		break;
	case sizeof(Atm_addr_spans):
		host_atm.address_format = T_ATM_SPANS_ADDR;
		host_atm.address_length = sizeof(Atm_addr_spans);
		break;
	default:
		fprintf(stderr, "%s: Invalid ATM address\n", prog);
		exit(1);
	}

	/*
	 * Build IOCTL request
	 */
	apr.aar_opcode = AIOCS_ADD_ARP;
	apr.aar_arp_dst.sa = host_ip.sa;
	ATM_ADDR_COPY(&host_atm, &apr.aar_arp_addr);
	apr.aar_arp_origin = ARP_ORIG_PERM;

	/*
	 * Tell the kernel to add the ARP table entry
	 */
	s = socket(AF_ATM, SOCK_DGRAM, 0);
	if (s < 0) {
		sock_error(errno);
	}
	if (ioctl(s, AIOCADD, (caddr_t)&apr) < 0) {
		fprintf(stderr, "%s: ", prog);
		switch (errno) {
		case EINVAL:
			fprintf(stderr, "Invalid parameter\n");
			break;
		case EPERM:
			fprintf(stderr, "Must be super user to use add subcommand\n");
			break;
		case EADDRNOTAVAIL:
			fprintf(stderr, "IP address not valid for interface\n");
			break;
		default:
			perror("ioctl (AIOCADD) add");
			break;
		}
		exit(1);
	}
	(void)close(s);
}


/*
 * Process PVC delete command
 * 
 * Command formats: 
 *	atm delete pvc <interface_name> <vpi> <vci>
 *
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
pvc_dlt(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	struct atmdelreq	apr;

	/*
	 * Set opcode
	 */
	apr.adr_opcode = AIOCS_DEL_PVC;

	/*
	 * Complete request by calling subroutine
	 */
	vcc_dlt(argc, argv, cmdp, &apr);
}


/*
 * Process SVC delete command
 * 
 * Command formats: 
 *	atm delete svc <interface_name> <vpi> <vci>
 *
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
svc_dlt(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	struct atmdelreq	apr;

	/*
	 * Set opcode
	 */
	apr.adr_opcode = AIOCS_DEL_SVC;

	/*
	 * Complete request by calling subroutine
	 */
	vcc_dlt(argc, argv, cmdp, &apr);
}


/*
 * Complete an SVC or PVC delete command
 * 
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *	apr	pointer to ATM delete IOCTL structure
 *
 * Returns:
 *	none
 *
 */
void
vcc_dlt(argc, argv, cmdp, apr)
	int			argc;
	char			**argv;
	struct cmd		*cmdp;
	struct atmdelreq	*apr;
{
	char	*cp;
	long	v;
	int	s;

	/*
	 * Validate interface name
	 */
	if (strlen(argv[0]) > sizeof(apr->adr_pvc_intf) - 1) {
		fprintf(stderr, "%s: Illegal interface name\n", prog);
		exit(1);
	}
	strcpy(apr->adr_pvc_intf, argv[0]);
	argc--; argv++;

	/*
	 * Validate vpi/vci values
	 */
	v = strtol(argv[0], &cp, 0);
	if ((*cp != '\0') || (v < 0) || (v >= 1 << 8)) {
		fprintf(stderr, "%s: Invalid VPI value\n", prog);
		exit(1);
	}
	apr->adr_pvc_vpi = (u_short) v;
	argc--; argv++;

	v = strtol(argv[0], &cp, 0);
	if ((*cp != '\0') || (v < MIN_VCI) || (v >= 1 << 16)) {
		fprintf(stderr, "%s: Invalid VCI value\n", prog);
		exit(1);
	}
	apr->adr_pvc_vci = (u_short) v;
	argc--; argv++;

	/*
	 * Tell the kernel to delete the VCC
	 */
	s = socket(AF_ATM, SOCK_DGRAM, 0);
	if (s < 0) {
		sock_error(errno);
	}
	if (ioctl(s, AIOCDEL, (caddr_t)apr) < 0) {
		fprintf(stderr, "%s: ", prog);
		switch (errno) {
		case EINVAL:
			fprintf(stderr, "Invalid parameter\n");
			break;
		case ENOENT:
			fprintf(stderr, "VCC not found\n");
			break;
		case EALREADY:
			fprintf(stderr, "VCC already being closed\n");
			break;
		case ENXIO:
			fprintf(stderr, "%s is not an ATM device\n",
					apr->adr_pvc_intf);
			break;
		case EPERM:
			fprintf(stderr, "Must be super user to use delete subcommand\n");
			break;
		default:
			perror("ioctl (AIOCDEL) delete");
			break;
		}
		exit(1);
	}
	(void)close(s);
}


/*
 * Process ARP delete command
 * 
 * Command formats: 
 *	atm delete arp <IP addr>
 *
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
arp_dlt(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	int	s;
	struct atmdelreq	apr;
	struct sockaddr_in	*sin;
	union {
		struct sockaddr_in	sin;
		struct sockaddr		sa;
	} host_addr;

	/*
	 * Set opcode
	 */
	UM_ZERO(&apr, sizeof(apr));
	apr.adr_opcode = AIOCS_DEL_ARP;

	/*
	 * Get network interface name if one is present
	 */
	if (argc == 2) {
		check_netif_name(argv[0]);
		strcpy(apr.adr_arp_intf, argv[0]);
		argc--; argv++;
	}

        /*
         * Get IP address of specified host name
         */
	UM_ZERO(&host_addr, sizeof(host_addr));
	host_addr.sa.sa_family = AF_INET;
	sin = get_ip_addr(argv[0]);
	host_addr.sin.sin_addr.s_addr = sin->sin_addr.s_addr;
	apr.adr_arp_dst.sa = host_addr.sa;

	/*
	 * Tell the kernel to delete the ARP table entry
	 */
	s = socket(AF_ATM, SOCK_DGRAM, 0);
	if (s < 0) {
		sock_error(errno);
	}
	if (ioctl(s, AIOCDEL, (caddr_t)&apr) < 0) {
		fprintf(stderr, "%s: ", prog);
		switch (errno) {
		case EINVAL:
			fprintf(stderr, "Invalid parameter\n");
			break;
		case EPERM:
			fprintf(stderr, "Must be super user to use delete subcommand\n");
			break;
		default:
			perror("ioctl (AIOCDEL) delete");
			break;
		}
		exit(1);
	}
	(void)close(s);
}


/*
 * Process help command
 * 
 * Arguments:
 *	argc	number of arguments to command
 *	argv	pointer to argument strings
 *	cmdp	pointer to command description 
 *
 * Returns:
 *	none
 *
 */
void
help(argc, argv, cmdp)
	int		argc;
	char		**argv;
	struct cmd	*cmdp;
{
	usage(cmds, "");
}
